cpp/src/arrow/flight/server.cc

```diff
@@ -248,6 +250,19 @@ class GrpcServerCallContext : public ServerCallContext {
     for (const auto& instance : middleware_) {
       instance->CallCompleted(status);
     }
+
+    // Set custom headers to map the exact Arrow status for clients
+    // who want it.
+    if (ARROW_PREDICT_FALSE(!status.ok()) && context_) {
```

Comment:
Should there be a `ToGrpcStatus(const Status&, grpc::ServerContext* context = nullptr)` instead?

Reply:
Yes, I've reworked it to keep the logic internal to internal.cc now.